ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriter.scala

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import akka.stream.scaladsl.{Keep, Sink, Source, SourceQueueWithComplete}
+import akka.stream.{Materializer, OverflowStrategy, QueueOfferResult}
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.DamlSubmissionBatch
+import com.daml.ledger.participant.state.kvutils.Envelope
+import com.daml.ledger.participant.state.v1.{ParticipantId, SubmissionResult}
+import com.digitalasset.ledger.api.health.HealthStatus
+import com.digitalasset.logging.{ContextualizedLogger, LoggingContext}
+import com.google.protobuf.ByteString
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+class BatchingLedgerWriter(
+    val writer: LedgerWriter,
+    val maxQueueSize: Int,
+    val maxBatchSizeBytes: Long,
+    val maxWaitDuration: FiniteDuration,
+    val maxParallelism: Int)(
+    implicit val materializer: Materializer,
+    implicit val logCtx: LoggingContext)
+    extends LedgerWriter {
+
+  implicit val executionContext: ExecutionContext = materializer.executionContext
+  private val logger = ContextualizedLogger.get(getClass)
+
+  /** Flag to mark whether the submission queue has been completed or not. Used for health status. */
+  private var queueAlive = true
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any")) /* Keep.left */
+  private val queue: SourceQueueWithComplete[DamlSubmissionBatch.CorrelatedSubmission] = {
+    // Sink to commit batches coming from the queue. We use foreachAsync to commit them in parallel.
+    // TODO(JM): Verify that foreachAsync actually works correctly and that we use sensible executionContext!
+    val commitSink: Sink[Seq[DamlSubmissionBatch.CorrelatedSubmission], _] =
+      Sink.foreachAsync(maxParallelism)(commitBatch)
+
+    val queue = Source
+      .queue(maxQueueSize, OverflowStrategy.dropNew)
+      .groupedWeightedWithin(maxBatchSizeBytes, maxWaitDuration)(
+        (cs: DamlSubmissionBatch.CorrelatedSubmission) => cs.getSubmission.size.toLong)
+      .toMat(commitSink)(Keep.left)
+      .run
+
+    // Watch for completion of the queue to mark the writer as unhealthy, so that upper layers can
+    // handle restarting.
+    // TODO(JM): Is there a neater way to accomplish this?
```

Comment:
Can't think of any

Reply:
Use a semaphore to signal if the source has been completed.

Reply:
Switched to using AtomicBoolean. I don't see how semaphore would help.

Reply:
AtomicBoolean is a good enough simple solution.